File: backend/services/workflow/graph_builder.py
# Blocker graph builder.
# Mirrors frontend/lib/graphBuilder.ts so the backend can return a graph
# in the shared schema: {"nodes": [...], "edges": [...]}.

from typing import Any

AT_LEAST_MODERATE = {"moderate", "medium_high", "high"}


def _node(id: str, label: str, status: str, owner: str) -> dict[str, str]:
    return {"id": id, "label": label, "status": status, "owner": owner}


def _level(categories: dict[str, Any], key: str) -> str | None:
    cat = categories.get(key) or {}
    return cat.get("level")


def build_graph(risk: dict[str, Any]) -> dict[str, list[dict[str, Any]]]:
    """Derive the blocker graph from the risk analysis section.

    `risk` is the shared-schema risk object:
      {"score": int, "level": str, "categories": {...}, "reasons": [...]}
    """
    categories = risk.get("categories") or {}
    overall = risk.get("level")

    nodes = [_node("opportunity_uploaded", "Opportunity uploaded", "clear", "student")]
    edges: list[dict[str, str]] = []
    signal_ids: list[str] = []

    if _level(categories, "paid_role") in AT_LEAST_MODERATE:
        nodes.append(_node("paid_role", "Paid role detected", "warning", "organizer"))
        edges.append({"from": "opportunity_uploaded", "to": "paid_role"})
        signal_ids.append("paid_role")

    if _level(categories, "work_authorization") in AT_LEAST_MODERATE:
        nodes.append(
            _node("work_auth", "Work authorization may be required", "blocked", "student")
        )
        parent = "paid_role" if "paid_role" in signal_ids else "opportunity_uploaded"
        edges.append({"from": parent, "to": "work_auth"})
        signal_ids.append("work_auth")

    if _level(categories, "citizenship") == "high":
        nodes.append(
            _node(
                "citizenship_restricted",
                "Citizenship restriction detected",
                "blocked",
                "system",
            )
        )
        edges.append({"from": "opportunity_uploaded", "to": "citizenship_restricted"})
        signal_ids.append("citizenship_restricted")

    if _level(categories, "ambiguity") == "high":
        nodes.append(
            _node(
                "eligibility_unclear",
                "International eligibility unclear",
                "blocked",
                "system",
            )
        )
        parents = [s for s in signal_ids if s != "eligibility_unclear"]
        if not parents:
            edges.append({"from": "opportunity_uploaded", "to": "eligibility_unclear"})
        else:
            edges.extend({"from": p, "to": "eligibility_unclear"} for p in parents)
        signal_ids.append("eligibility_unclear")

    verification_needed = overall in {"high", "medium_high"}
    leaves = [signal_ids[-1]] if signal_ids else ["opportunity_uploaded"]

    if verification_needed:
        nodes.extend(
            [
                _node("organizer_verify", "Organizer verification needed", "pending", "organizer"),
                _node(
                    "advisor_verify",
                    "Advisor/DSO verification recommended",
                    "pending",
                    "advisor",
                ),
                _node(
                    "decision",
                    "Do not assume eligibility until confirmed",
                    "blocked",
                    "student",
                ),
            ]
        )
        for leaf in leaves:
            edges.append({"from": leaf, "to": "organizer_verify"})
            edges.append({"from": leaf, "to": "advisor_verify"})
        edges.append({"from": "organizer_verify", "to": "decision"})
        edges.append({"from": "advisor_verify", "to": "decision"})
    else:
        if signal_ids:
            nodes.append(
                _node("decision", "Verify remaining questions, then proceed", "warning", "student")
            )
        else:
            nodes.append(
                _node(
                    "decision",
                    "No major blockers detected — safe to proceed",
                    "clear",
                    "student",
                )
            )
        edges.extend({"from": leaf, "to": "decision"} for leaf in leaves)

    return {"nodes": nodes, "edges": edges}

File: frontend/lib/graphBuilder.ts
import type {
  GraphEdge,
  GraphNode,
  PartialVisaLensAnalysis,
  RiskLevel,
} from "@/types/analysis";

/**
 * Graph builder for the blocker/dependency graph.
 *
 * Two jobs:
 *  1. resolveGraph(): use the backend graph if present, otherwise derive
 *     a graph from risk categories (fallback so the UI never renders empty).
 *  2. layoutGraph(): assign x/y positions with a simple layered
 *     (topological) layout so React Flow can render without dagre.
 */

export type BlockerGraphData = { nodes: GraphNode[]; edges: GraphEdge[] };

const AT_LEAST_MODERATE: RiskLevel[] = ["moderate", "medium_high", "high"];

function levelAtLeastModerate(level?: RiskLevel): boolean {
  return !!level && AT_LEAST_MODERATE.includes(level);
}

/**
 * Build a blocker graph from risk categories when the backend did not
 * send one. Mirrors backend/services/workflow/graph_builder.py.
 */
export function buildGraphFromRisk(
  analysis: PartialVisaLensAnalysis
): BlockerGraphData {
  const risk = analysis.risk ?? {};
  const categories = risk.categories;
  const overallLevel = risk.level;

  const nodes: GraphNode[] = [
    {
      id: "opportunity_uploaded",
      label: "Opportunity uploaded",
      status: "clear",
      owner: "student",
    },
  ];
  const edges: GraphEdge[] = [];

  // Risk-signal layer. Each detected signal hangs off the root.
  const signalIds: string[] = [];

  if (levelAtLeastModerate(categories?.paid_role?.level)) {
    nodes.push({
      id: "paid_role",
      label: "Paid role detected",
      status: "warning",
      owner: "organizer",
    });
    edges.push({ from: "opportunity_uploaded", to: "paid_role" });
    signalIds.push("paid_role");
  }

  if (levelAtLeastModerate(categories?.work_authorization?.level)) {
    nodes.push({
      id: "work_auth",
      label: "Work authorization may be required",
      status: "blocked",
      owner: "student",
    });
    // Work auth follows the paid-role signal when present, else the root.
    edges.push({
      from: signalIds.includes("paid_role") ? "paid_role" : "opportunity_uploaded",
      to: "work_auth",
    });
    signalIds.push("work_auth");
  }

  if (categories?.citizenship?.level === "high") {
    nodes.push({
      id: "citizenship_restricted",
      label: "Citizenship restriction detected",
      status: "blocked",
      owner: "system",
    });
    edges.push({ from: "opportunity_uploaded", to: "citizenship_restricted" });
    signalIds.push("citizenship_restricted");
  }

  if (categories?.ambiguity?.level === "high") {
    nodes.push({
      id: "eligibility_unclear",
      label: "International eligibility unclear",
      status: "blocked",
      owner: "system",
    });
    // Ambiguity is downstream of whichever signals exist.
    const parents = signalIds.filter((id) => id !== "eligibility_unclear");
    if (parents.length === 0) {
      edges.push({ from: "opportunity_uploaded", to: "eligibility_unclear" });
    } else {
      for (const p of parents) edges.push({ from: p, to: "eligibility_unclear" });
    }
    signalIds.push("eligibility_unclear");
  }

  // Verification layer for risky opportunities.
  const verificationNeeded =
    overallLevel === "high" || overallLevel === "medium_high";

  const leafSignals =
    signalIds.length > 0 ? [signalIds[signalIds.length - 1]] : ["opportunity_uploaded"];

  if (verificationNeeded) {
    nodes.push(
      {
        id: "organizer_verify",
        label: "Organizer verification needed",
        status: "pending",
        owner: "organizer",
      },
      {
        id: "advisor_verify",
        label: "Advisor/DSO verification recommended",
        status: "pending",
        owner: "advisor",
      },
      {
        id: "decision",
        label: "Do not assume eligibility until confirmed",
        status: "blocked",
        owner: "student",
      }
    );
    for (const leaf of leafSignals) {
      edges.push({ from: leaf, to: "organizer_verify" });
      edges.push({ from: leaf, to: "advisor_verify" });
    }
    edges.push({ from: "organizer_verify", to: "decision" });
    edges.push({ from: "advisor_verify", to: "decision" });
  } else {
    nodes.push({
      id: "decision",
      label:
        signalIds.length > 0
          ? "Verify remaining questions, then proceed"
          : "No major blockers detected — safe to proceed",
      status: signalIds.length > 0 ? "warning" : "clear",
      owner: "student",
    });
    for (const leaf of leafSignals) edges.push({ from: leaf, to: "decision" });
  }

  return { nodes, edges };
}

/**
 * Prefer the backend graph; fall back to risk-derived graph; final
 * fallback is a minimal two-node graph so the canvas is never empty.
 */
export function resolveGraph(
  analysis: PartialVisaLensAnalysis
): BlockerGraphData {
  const g = analysis.graph;
  if (g?.nodes && g.nodes.length > 0) {
    return { nodes: g.nodes, edges: g.edges ?? [] };
  }
  if (analysis.risk?.categories) {
    return buildGraphFromRisk(analysis);
  }
  return {
    nodes: [
      {
        id: "opportunity_uploaded",
        label: "Opportunity uploaded",
        status: "clear",
        owner: "student",
      },
      {
        id: "awaiting_analysis",
        label: "Awaiting analysis",
        status: "pending",
        owner: "system",
      },
    ],
    edges: [{ from: "opportunity_uploaded", to: "awaiting_analysis" }],
  };
}

// ---------------------------------------------------------------------------
// Layout
// ---------------------------------------------------------------------------

export type PositionedNode = GraphNode & { x: number; y: number };

const NODE_WIDTH = 240;
const X_GAP = 60;
const Y_GAP = 110;

/**
 * Simple layered layout: level = longest path from a root node.
 * Nodes in the same level are spread horizontally and centered.
 * Good enough for our ~6-9 node workflow graphs, no dagre needed.
 */
export function layoutGraph(graph: BlockerGraphData): PositionedNode[] {
  const { nodes, edges } = graph;
  const ids = new Set(nodes.map((n) => n.id));
  const validEdges = edges.filter((e) => ids.has(e.from) && ids.has(e.to));

  // Longest-path level assignment (graphs are small; O(V*E) is fine).
  const level = new Map<string, number>();
  for (const n of nodes) level.set(n.id, 0);
  for (let i = 0; i < nodes.length; i++) {
    let changed = false;
    for (const e of validEdges) {
      const next = (level.get(e.from) ?? 0) + 1;
      if (next > (level.get(e.to) ?? 0)) {
        level.set(e.to, next);
        changed = true;
      }
    }
    if (!changed) break;
  }

  // Group by level, then center each row.
  const rows = new Map<number, GraphNode[]>();
  for (const n of nodes) {
    const l = level.get(n.id) ?? 0;
    if (!rows.has(l)) rows.set(l, []);
    rows.get(l)!.push(n);
  }

  const maxRowWidth = Math.max(
    ...Array.from(rows.values()).map(
      (row) => row.length * NODE_WIDTH + (row.length - 1) * X_GAP
    )
  );

  const positioned: PositionedNode[] = [];
  for (const [l, row] of Array.from(rows.entries()).sort((a, b) => a[0] - b[0])) {
    const rowWidth = row.length * NODE_WIDTH + (row.length - 1) * X_GAP;
    const startX = (maxRowWidth - rowWidth) / 2;
    row.forEach((n, i) => {
      positioned.push({
        ...n,
        x: startX + i * (NODE_WIDTH + X_GAP),
        y: l * Y_GAP,
      });
    });
  }
  return positioned;
}
